### PR TITLE
Add type hints to helpers and update DBF extraction tests

### PR DESCRIPTION
**Pull request by zaneselvans**
# PR Overview

This is just me being a little bit OCD.

* Added type hints to all remaining functions in pudl.helpers that didn't have them.
* Added schema compatability checks for FERC 6 and 60 DBF derived SQLite DBs.
* Use a context manager when dropping tables in `pudl.helpers.drop_tables()`

# PR Checklist

- [ ] Merge the most recent version of the branch you are merging into (probably `dev`).
- [ ] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [ ] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [ ] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

### CODE REVIEW

File: src/pudl/helpers.py [hunk 1]
```diff
@@ -287,45 +304,44 @@ def oob_to_nan(df, cols, lb=None, ub=None):
...
-        dir_path (path-like): path to the directory that you are trying to
-            clean and prepare.
-        clobber (bool): If True and dir_path exists, it will be removed and
-            replaced with a new, empty directory.
+        dir_path: path to the directory that you are trying to clean and prepare.
+        clobber: If True and dir_path exists, it will be removed and replaced with a
+            new, empty directory.
 
     Raises:
         FileExistsError: if a file or directory already exists at dir_path.
```
Comment (e-belfer) [inline src/pudl/helpers.py:347]: For consistency, scratch the pathlib.Path type in the "returns" portion of the description as well.

File: src/pudl/helpers.py [hunk 2]
```diff
@@ -688,26 +705,31 @@ def simplify_strings(df, columns):
...
+) -> pd.Series:
     """Clean up the strings in a single column/Series.
 
     Args:
-        col (pandas.Series): A pandas Series, typically a single column of a
+        col: A pandas Series, typically a single column of a
             dataframe, containing the freeform strings that are to be cleaned.
-        str_map (dict): A dictionary of lists of strings, in which the keys are
+        str_map: A dictionary of lists of strings, in which the keys are
             the simplified canonical strings, witch which each string found in
```
Comment (e-belfer) [inline src/pudl/helpers.py:720]: "with which"

File: src/pudl/helpers.py [hunk 3]
```diff
@@ -948,20 +973,17 @@ def convert_to_date(
...
-        month_col (str): the name of the month column in the original table.
+        df: dataframe to convert
+        date_col: the name of the column you want in the output.
+        year_col: the name of the year column in the original table.
+        month_col: the name of the month column in the original table.
         day_col: the name of the day column in the original table.
-        month_value (int): generated month if no month exists.
-        day_value (int): generated day if no month exists.
+        month_value: generated month if no month exists.
+        day_value: generated day if no month exists.
```
Comment (e-belfer) [inline src/pudl/helpers.py:982]: If no month or if no day? Based on code seems to be day.

File: src/pudl/helpers.py [hunk 4]
```diff
@@ -1049,34 +1066,15 @@ def drop_tables(engine: sa.engine.Engine, clobber: bool = False):
...
-        raise AssertionError(
-            f"You are attempting to drop your database at {engine} while clobber is set to {clobber}"
-        )
-    md.drop_all(engine)
-    conn = engine.connect()
-    conn.exec_driver_sql("VACUUM")
-    conn.close()
-
-
-def merge_dicts(list_of_dicts):
```
Comment (e-belfer) [inline src/pudl/helpers.py:1064]: Even if unused currently, can we imagine a use for this function? If so I'd suggest just keeping it.

**Review by e-belfer — APPROVED** [review]
Some other changes suggested while you're at it, and one non-blocking question. Otherwise looks good, presuming the code runs as expected.

File: src/pudl/helpers.py (hunk 1 above)
Comment (zaneselvans) [reply src/pudl/helpers.py:347]: Oops! Missed this one. Updated.

File: src/pudl/helpers.py (hunk 4 above)
Comment (zaneselvans) [reply src/pudl/helpers.py:1064]: I can imagine it being useful, but whatever the actual use case was that we had seems to have been refactored away, and it's almost a one-liner... I feel like there should be an easy way to do this with a dictionary comprehension but it's not coming to mind immediately.

Comment (zaneselvans) [reply src/pudl/helpers.py:1064]: Ah okay, as per Alex's suggestion

```py
{k: v for d in list_of_dicts for k, v in d.items()}
```

**Pull request merged**